python_ta/checkers/possibly_undefined_checker.py

```diff
@@ -0,0 +1,119 @@
+"""checker for variables that might not be defined in the program.
+"""
+from typing import Union
+import astroid
+from pylint.interfaces import IAstroidChecker
+from pylint.checkers import BaseChecker
+from pylint.checkers.utils import check_messages
+from python_ta.cfg.graph import CFGBlock, ControlFlowGraph
+from typing import Set
+
+
+class PossiblyUndefinedChecker(BaseChecker):
+
+    __implements__ = IAstroidChecker
+    # name is the same as file name but without _checker part
+    name = 'possibly_undefined'
+    # use dashes for connecting words in message symbol
+    msgs = {'E9969': ('This variable might not be defined when this statement is executed.',
+                      'possibly-undefined',
+                      'Reported when a statement uses a variable that might not be assigned.'),
+           }
+
+    # this is important so that your checker is executed before others
+    priority = -1
+
+    def __init__(self, linter=None):
+        super().__init__(linter=linter)
+        self._possibly_undefined: Set[astroid.Name] = set()
+
+    @check_messages('possibly-undefined')
+    def visit_name(self, node):
+        """Adds message if there exists a path from the start block to node where
+        a variable used by node might not be defined."""
+        if node in self._possibly_undefined:
+            self.add_message('possibly-undefined', node=node)
+
+    def visit_module(self, node):
+        self._analyze(node)
+
+    def visit_functiondef(self, node):
+        self._analyze(node)
+
+    def _analyze(self, node: Union[astroid.Module, astroid.FunctionDef]) -> None:
+        """Runs the data flow algorithm on a `Module` or `Function` CFG, which in turn
+        appends `Name` nodes to `self.possibly_undefined` if it might not be defined.
+
+        Data flow algorithms retrieved from:
+        https://www.seas.harvard.edu/courses/cs252/2011sp/slides/Lec02-Dataflow.pdf#page=31
+        """
+        out_facts = {}
+        blocks = list(node.cfg.get_blocks_postorder())
+        blocks.reverse()
+
+        all_assigns = self._get_assigns(node)
+        for block in blocks:
+            out_facts[block] = {}
+            out_facts[block] = all_assigns.copy()
```

Comment:
I'd like to double-check this---isn't it sufficient to initialize `out_facts[block]` to be empty? It seems like the main work should be done in the `_transfer` to accumulate the actual `AssignName` nodes that populate the out facts.

Either way, lines 56 and 57 conflict, so one of them should be removed.

Reply:
According to the reference, the `outs` must be set to the top of the lattice (i.e all the local variables).

A case where `outs` being empty is not sufficient is when there's a back-edge.
 
```
y = 0
while y > 10:
    x = 0
```

![cfg_misc](https://user-images.githubusercontent.com/18351436/64042080-16ec5800-cb2f-11e9-8f75-d7ab9e594955.jpg)

Since the `join` operation is set `intersection`, If `outs` wasn't set to all the local variables then `y` won't be definitely assigned at the loop condition block.

Reply:
Alright, that makes sense to me. Thanks!